File: Structural_Patterns/Flyweight/cpp/Flyweight.cpp
/*!
 *  \file       Flyweight.cpp
 *  \brief      
 *  
 */


#include "Flyweight.hpp"

#include "SharedState.hpp"
#include "UniqueState.hpp"

namespace dive_into_design_patterns
{
    Flyweight::Flyweight(const SharedState *shared_state) : m_shared_state{new SharedState(*shared_state)}
    {
    }

    Flyweight::Flyweight(const Flyweight &other) : m_shared_state{new SharedState(*other.m_shared_state)}
    {
    }

    Flyweight::~Flyweight()
    {
        if (m_shared_state)
        {
            delete m_shared_state;
            m_shared_state = nullptr;
        }
    }

    SharedState* Flyweight::getSharedState() const
    {
        return m_shared_state;
    }

    void Flyweight::operation(const UniqueState &unique_state) const
    {
        std::cout << "Flyweight: Displaying shared (" << *m_shared_state 
                  << ") and unique (" << unique_state << ") state."
                  << std::endl;
    }
}

File: Structural_Patterns/Flyweight/python/Flyweight.py
# File name     : Flyweight.py
# Description   :
#
#

import json

class Flyweight:
    '''
    The Flyweight stores a common portion of the state (also called intrinsic state) that belongs to multiple 
    real business entities. The Flyweight accepts the rest of the state (extrinsic state, unique for each entity)
    via its method parameters.
    '''
    def __init__(self, shared_state: str):
        self._shared_state = shared_state

    def operation(self, unique_state: str):
        s = json.dumps(self._shared_state)
        u = json.dumps(unique_state)
        print(f"Flyweight: Displaying shared ({s}) and unique ({u}) state.", end="")